sklearn/preprocessing/target.py

```diff
@@ -0,0 +1,242 @@
+            y_2d = y.reshape(-1, 1)
+        else:
+            y_2d = y
+        self._fit_transformer(y_2d, sample_weight)
+        self.regressor_ = clone(self.regressor)
+        if sample_weight is not None:
+            self.regressor_.fit(X, self.transformer_.fit_transform(y_2d),
+                                sample_weight=sample_weight)
+        else:
+            self.regressor_.fit(X, self.transformer_.fit_transform(y_2d))
+        return self
+
+    def predict(self, X):
+        """Predict using the base regressor, applying inverse.
+
+        The regressor is used to predict and the ``inverse_func`` or
+        ``inverse_transform`` is applied before returning the prediction.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features)
+            Samples.
+
+        Returns
+        -------
+        y_hat : array, shape = (n_samples,)
+            Predicted values.
+
+        """
+        check_is_fitted(self, "regressor_")
+        pred = self.transformer_.inverse_transform(self.regressor_.predict(X))
+        # if y is not a multi-output, it should be ravel
+        if self.y_ndim_ == 1 and self.func is None:
+            return pred.ravel()
+        else:
+            return pred
+
+    def score(self, X, y, sample_weight=None):
+        """Returns the coefficient R^2 of the prediction in the original space.
+
+        The coefficient R^2 is defined as (1 - u/v), where u is the regression
+        sum of squares ((y_true - y_pred) ** 2).sum() and v is the residual sum
+        of squares ((y_true - y_true.mean()) ** 2).sum().  Best possible score
+        is 1.0 and it can be negative (because the model can be arbitrarily
+        worse). A constant model that always predicts the expected value of y,
+        disregarding the input features, would get a R^2 score of 0.0. Note
+        that the score is computed in the original space using the
+        ``inverse_transform`` of ``transformer_``.
+
+        Parameters
+        ----------
+        X : array-like, shape = (n_samples, n_features)
+            Test samples.
+
+        y : array-like, shape = (n_samples) or (n_samples, n_outputs)
+            True values for X.
+
+        sample_weight : array-like, shape = [n_samples], optional
+            Sample weights.
+
+        Returns
+        -------
+        score : float
+            R^2 of self.predict(X) wrt. y.
+
+        """
+
+        check_is_fitted(self, "regressor_")
+        if not is_regressor(self.regressor_):
+            if not hasattr(self.regressor_, "_estimator_type"):
+                err = "regressor has declared no _estimator_type."
+            else:
+                err = "regressor has _estimator_type {}".format(
+                    self.regressor_._estimator_type)
+            raise NotImplementedError("TransformedTargetRegressor should be a"
+                                      " regressor. This " + err)
```

Comment:
"This" -> self.regressor_.__name__ or whatever works?

If the user is trying to use a custom defined regressor, we should give an easy fix here, e.g. "inherit from `sklearn.base.RegressorMixin`" (assuming that fixes it?)